src/graphics/reproject-texture.js

```diff
@@ -281,74 +306,97 @@ const generateGGXSamples = (numSamples, specularPower, sourceTotalPixels) => {
 };
 
 // pack float samples data into an rgba8 texture
-const packSamplesTex = (device, name, samples) => {
-    const numSamples = samples.length;
-
-    const w = Math.min(numSamples, 512);
-    const h = Math.ceil(numSamples / w);
-    const data = new Uint8Array(w * h * 4);
-
-    // normalize float data and pack into rgba8
-    let off = 0;
-    for (let i = 0; i < numSamples; ++i) {
-        packFloat32ToRGBA8(samples[i * 4 + 0] * 0.5 + 0.5, data, off + 0);
-        packFloat32ToRGBA8(samples[i * 4 + 1] * 0.5 + 0.5, data, off + 4);
-        packFloat32ToRGBA8(samples[i * 4 + 2] * 0.5 + 0.5, data, off + 8);
-        packFloat32ToRGBA8(samples[i * 4 + 3] / 8, data, off + 12);
-        off += 16;
-    }
-
+const createSamplesTex = (device, name, samples) => {
+    const packedSamples = packSamples(samples);
     return new Texture(device, {
         name: name,
-        width: w,
-        height: h,
+        width: packedSamples.width,
+        height: packedSamples.height,
         mipmaps: false,
         minFilter: FILTER_NEAREST,
         magFilter: FILTER_NEAREST,
-        levels: [data]
+        levels: [packedSamples.data]
     });
 };
 
-// cache of float sample data packed into rgba8 textures
-const samplesTexCache = { };
+// simple cache storing key->value
+// missFunc is called if the key is not present
+class SimpleCache {
+    constructor() {
+        this.map = new Map();
+    }
 
-const generateLambertSamplesTex = (device, numSamples, sourceTotalPixels) => {
-    const key = `lambert-samples-${numSamples}-${sourceTotalPixels}`;
+    get(key, missFunc) {
+        if (!this.map.has(key)) {
+            const result = missFunc();
+            this.map.set(key, result);
+            return result;
+        }
+        return this.map.get(key);
+    }
+
+    clear() {
+        this.map.clear();
+    }
+}
 
-    if (samplesTexCache.hasOwnProperty(key)) {
-        return samplesTexCache[key];
+// per-device cache
+class DeviceCache {
+    constructor() {
+        this.cache = new SimpleCache();
     }
 
-    const samples = generateLambertSamples(numSamples, sourceTotalPixels);
-    const samplesTex = packSamplesTex(device, key, samples);
-    samplesTexCache[key] = samplesTex;
-    return samplesTex;
+    get(device, key, missFunc) {
+        return this.cache.get(device, () => {
+            device.on('destroy', () => {
+                this.cache.delete(device);
```

Comment:
should this loop over textures in the cache and call destroy on them too?

Reply:
I wondered that too, but don't think it's necessary. Calling `destroy()` on the texture is necessary to clear up the graphics device-side objects, but that will happen anyway since the device is busy being destroyed. The other reason to invoke destroy on texture is to clear references to the underlying `levels` data. In our case this isn't necessary either since it's cached.

That said, perhaps we should destroy them anyway? On the one hand I don't like adding unnecessary code 'just in case', but on the other hand it might be good to "set a precedent".

What do you think?

Reply:
I prefer to destroy all resources like this if possible, as this makes tracking application side leaks easier - for example by printing the list of textures left over in the device.destroy